### ISSUE TITLE
Extending the ndarray protocol to support xnd and in particular libndtypes

### ISSUE THREAD

**Issue by teoliphant** [hameerabbasi/sparse-ndarray-protocols#1]
For a while we (meaning Stefan Krah and I and a few other folks from time to time) have been working steadily on an updated type description to improve PEP 3118.    This work is nearing MVP in plures/xnd

At it's core is a type-description language that allows to describe all kinds of data that might be in a typed container which answers the fundamental problem which was at the heart of why PEP 3118 defaulted to "struct-strings" as a type-description language instead of something more comprehensive.  

The [Plures](https://github.com/plures) organization on GitHub holds all of these modules that effectively are "refactorings" and generalizations of NumPy components:
dtype => ilbndtypes (C-library), ndtypes (Python-module)
ufunc object ==> libgumath (C-library), gumath (Python-module)
ndarray object ==> libxnd (C-library), xnd (Python-module)

I'd propose extending the sparse ndarray protocol here to (optionally) support libndtypes type-system (datashape) so that xnd can participate in the sparse-array protocol. 

The limitations of NumPy's dtype system are known as are the limitations of PEP 3118 struct-string type system.  This is exactly what xnd is trying to solve.

**Comment by hameerabbasi** [comment]
If I understand correctly, you're proposing that sparse arrays support `ndtypes` types in addition to Numpy `dtype`s using `xnd` data containers instead of `ndarray` containers? This seems like a decent idea in principle. I have a question: Can Numpy Arrays currently support `ndtypes` dtypes?

The point of this proposed specification is specifically to mimic parts of the `ndarray` interface with sparse arrays without mimicking the memory layout. This proposal could (in principle) make sparse arrays incompatible with `ndarray`s if the answer to the above question is "no".

However, if I understand correctly, your long-term goal is to get these cleaned-up and generalized libraries upstreamed into Numpy?

**Comment by hameerabbasi** [comment]
However, some of the things I see here are indeed useful. You might want to coordinate with numpy/numpy_stubs. Of particular interest is numpy/numpy_stubs#5

**Comment by teoliphant** [comment]
NumPy would need to be re-factored and built on-top of a more generalized type-system like ndtypes, which is indeed a possibility, but not necessary for sparse to use them.   

NumPy Arrays cannot support ndtypes without extensive re-factoring and work.  This is why xnd exists.  Long-term NumPy could be built on xnd libraries if that is useful to enough people.

**Comment by hameerabbasi** [comment]
Does `xnd` follow the Numpy array protocol? If so, this specification implicitly supports this. We could, of course, make that more clear.

**Comment by teoliphant** [comment]
Your docs are clear that you support the array protocol.  As the author of the array protocol, I understand its uses and its limitations.

Xnd does support the array protocol as far as it can.  However, this is not enough as the array protocol doesn't have a complete type system.   This is one of the problems that xnd solves.

libndtypes > numpy.dtypes (so if you support libndtypes you can easily also support numpy.dtypes but the other way around is not true).

xnd arrays > ndarrays (xnd arrays support more kinds of arrays than NumPy can). 

As xnd matures, it will easy to see how NumPy can be adapted to use it.  However, iteratively changing NumPy to support the features of xnd is much, much harder.

**Comment by hameerabbasi** [comment]
If xnd arrays are strict supersets in terms of API support, then yes, it makes sense to support this type system. You're welcome to send a PR.

**Comment by teoliphant** [comment]
Great!  Thanks.   It might be a few weeks.

**Comment by eric-wieser** [comment]
> libndtypes > numpy.dtypes

Last time I looked, libndtypes did not support overlapping fields, while `dtype` does. Is this still the case?

**Comment by skrah** [comment]
I added a union type today:

https://github.com/plures/ndtypes/commit/4d810d0c4d54c81a7136f313f0ae6623853d574a

It is like a C union, so e.g. `{a: [int64, complex128], b: [string, float64]}` is a type of a struct that contains two union fields.

Would that be sufficient? The NumPy docs (https://www.numpy.org/devdocs/reference/arrays.dtypes.html) seem to indicate that overlapping fields are not encouraged for struct fields:

"This form also makes it possible to specify struct dtypes with overlapping fields, functioning like the ‘union’ type in C. This usage is discouraged, however, and the union mechanism is preferred."

So I'm not sure whether the union type would meet current best practices in NumPy or not.

**Comment by skrah** [comment]
And of course unions outside a struct are also possible:

`2 * 3 * [int64, complex128]`

**Comment by eric-wieser** [comment]
Those numpy docs are unclear - I think the "union mechanism" they claim is preferred is not actually documented anyway. What I believe they are referring to is:

```python
np.dtype(dict(
    formats=[np.int32, np.int32, np.int32],
    names=['first', 'middle', 'last'],
    offsets=[0, 2, 4],
    itemsize=8
))
```

Which creates partially-overlapping fields.

**Comment by hameerabbasi** [comment]
I'm hard pressed to find a use-case for this outside of very esoteric bit-twiddling. I think one aim was to keep libndtypes code clean and easy to read, and I think this implies that we shouldn't make the code a mess for those 0.1% use-cases.

I'd love to be proven wrong, and if there is a popular or impactful use-case, I'd be +1 on this.

**Comment by eric-wieser** [comment]
The example I gave above was more to prove that `ndtypes` is not a superset of `np.dtype`, not necessarily to show that `ndtypes` is inferior.

There are two places where I think this difference does matter, unlike the somewhat contrived case above:

1. Strides that aren't a multiple of steps
    > For NumPy compatibility ndtypes displays strides (amount of bytes to skip). **Internally, libndtypes uses steps (amount of indices to skip).**

    This seems to rule out views on structured arrays. Consider the following:
    ```python
    dt = np.dtype([('a', np.int32), ('b', np.int16)])
    arr = np.zeros(4, dt)
    arr_a = arr['a']
    arr_a.itemsize  # 4
    arr_a.strides  # (6,)
    ```
    My reading of the ndtypes documentation is that it is unable to represent `arr_a`

2. Fields that are out of order

    Numpy is moving in the direction of allowing views on structured arrays that permute the fields, as it provides a powerful way to support field-swapping assignment between arrays. This results in non-overlapping-but-out-of-order dtypes like:

    ```
    dt = np.dtype(dict(
        formats=[np.int32, np.int32],
        names=['y', 'x'],
        offsets=[4, 0],
        itemsize=8
   ))
   ```

   From what I've gleaned from the ndtypes docs so far, this is also not supported.

Even if `ndtypes` has a better internal design in some way, it's going to be very difficult to migrate to it from `np.dtype` without it being a complete superset.

**Comment by eric-wieser** [comment]
> And of course unions outside a struct are also possible: `2 * 3 * [int64, complex128]`

Have you considered spelling that `2 * 3 * Union[int64, complex128]` for consistency with the builtin `typing.Union`?

**Comment by eric-wieser** [comment]
> have been working steadily on an updated type description to improve PEP 3118

It is worth remembering that ndtypes _does_ succeed at being a superset of PEP3118, which is definitely welcome

**Comment by skrah** [comment]
Thanks for the examples!
 
>     1. Strides that aren't a multiple of steps
>        > For NumPy compatibility ndtypes displays strides (amount of bytes to skip). **Internally, libndtypes uses steps (amount of indices to skip).**
>        
>        
>        This seems to rule out views on structured arrays. Consider the following:
>        ```python
>        dt = np.dtype([('a', np.int32), ('b', np.int16)])
>        arr = np.zeros(4, dt)
>        arr_a = arr['a']
>        arr_a.itemsize  # 4
>        arr_a.strides  # (6,)
>        ``` 

It is indeed currently not possible.  Indexing through a record is:

```
>>> from xnd import *
>>> x = xnd.empty("4 * {a: int32, b: int16}")
>>> x[1, 0]
xnd(0, type='int32')
```

The reason for using steps is that `xnd` computes array accesses in terms of the linear index, which is frozen for views (just like the `ptr` is adjusted in PEP-3118).

The linear index is only applied to the data pointer at `ndim==0`.

`xnd` has offsets-based ragged arrays (same as Arrow list), which naturally use the linear index and not strides. Additionally, *any* arbitrarily nested type can contain optional values:

```
>>> x = xnd({'a': [(3.0j, "foo"), (4.0+10j, None)], 'b': "bar"})
>>> x
xnd({'a': [(3j, 'foo'), ((4+10j), None)], 'b': 'bar'}, type='{a : 2 * (complex128, ?string), b : string}')
>>> x['a', 1, 1]
xnd(None, type='?string')
```

The optional values use bitmaps, whose structure mirrors the type tree.   Bitmaps also need the linear index in order to address the validity bit.

Converting between strides and the index is not pleasant, since a) one always needs the start of the data and b) types like the empty tuple "()" have itemsize 0 (so one has to check before division).

One way around the situation would be to have a `type->logical_datasize` member, so it isn't impossible.


>     2. Fields that are out of order
>        Numpy is moving in the direction of allowing views on structured arrays that permute the fields, as it provides a powerful way to support field-swapping assignment between arrays. This results in non-overlapping-but-out-of-order dtypes like:
>        ```
>        dt = np.dtype(dict(
>            formats=[np.int32, np.int32],
>            names=['y', 'x'],
>            offsets=[4, 0],
>            itemsize=8
>        ))
>        ```
>        From what I've gleaned from the ndtypes docs so far, this is also not supported.

No, it's not supported.  Exactly the `pack` and `align` attributes from `gcc` are supported.

```
>>> from ndtypes import ndt
>>> t = ndt("{first: int32, middle: int32 |pack=2|, last: int32 |pack=2|}")
```

But `gcc` ignores packing that encroaches on the preceding type's territory.  We could loosen that requirement.

We could also allow something like:

```
>>> from ndtypes import ndt
>>> t = ndt("{x: int32 |offset=4|, y: int32 |offset=0|}")
```

**Comment by skrah** [comment]
> > And of course unions outside a struct are also possible: `2 * 3 * [int64, complex128]`
> 
> Have you considered spelling that `2 * 3 * Union[int64, complex128]` for consistency with the builtin `typing.Union`?

Yes, I've considered `union[]`.  Type signatures can get long though, so saving space is always a consideration. I'll think about it.

**Comment by eric-wieser** [comment]
> It is indeed currently not possible. Indexing through a record is:

So to be clear - there is no way to create a view over all the `a` fields in that struct, and you have to access them individually? That sounds like an important feature gap.

> No, it's not supported. Exactly the pack and align attributes from gcc are supported.

Actually, having solved this problem in C recently - I think you might be able to represent:
```
dt = np.dtype(dict(
       formats=[np.int32, np.int32],
       names=['y', 'x'],
       offsets=[4, 0],
       itemsize=8
   ))
```
in your language as:
```
[{_pad0: uint8 * 4, y: uint32}, {x: uint32, _pad1: uint8 * 4}
```

Support for unnamed struct fields would make that less awkward.

**Comment by eric-wieser** [comment]
> Exactly the pack and align attributes from gcc are supported.

I've never heard of the `pack(N)` attribute on fields, only the `packed` attribute on types. Am I missing something?

**Comment by skrah** [comment]
On Wed, Nov 21, 2018 at 03:22:25PM -0800, Eric Wieser wrote:
> I've never heard of the `pack(N)` attribute on fields, only the `packed` attribute on types. Am I missing something?

pack=2 is a libndtypes shorthand for __attribute__((packed, aligned(2))) to save space.

**Comment by skrah** [comment]
> > It is indeed currently not possible. Indexing through a record is:
> 
> So to be clear - there is no way to create a view over all the `a` fields in that struct, and you have to access them individually? That sounds like an important feature gap.

Indeed, there is no way to create that view, but slicing should be added to records, so perhaps `arr['a']` could be spelled `arr[:, 'a']`.

With columnar data types being all the rage, what *is* possible of course is to use this from the start:

```
>>> x = xnd.empty("{a: 4 * int32, b: 4 * int16}")
>>> x
xnd({'a': [0, 0, 0, 0], 'b': [0, 0, 0, 0]}, type='{a : 4 * int32, b : 4 * int16}')
>>> x['a']
xnd([0, 0, 0, 0], type='4 * int32')
>>> 
```
